File: python/DecisionTree.py
from itertools import *
import pandas as pd
from numpy import *
import random
from math import sqrt

class TreeNode(object):
    def __init__(self, feat, val, mc, left = 0, right = 0):
        self.feature = feat
        self.value = val
        self.maxClass = mc
        self.left = left
        self.right = right


class BTree(object):
    def __init__(self, root = 0):
        self.root = root
        
    def is_empty(self):
        if self.root is 0:
            return True
        else:
            return False
        
    def preOrder(self, treenode):
        if isinstance(treenode, str) or isinstance(treenode, int):
            return treenode
        print treenode.feature, treenode.value, treenode.maxClass
        self.preOrder(treenode.left)
        self.preOrder(treenode.right)
        
    def inOrder(self, treenode):
        if isinstance(treenode, str) or isinstance(treenode, int):
            return treenode
        self.inOrder(treenode.left)
        print treenode.feature, treenode.value, treenode.maxClass
        self.inOrder(treenode.right)
        
    def postOrder(self, treenode):
        if isinstance(treenode, str) or isinstance(treenode, int):
            return treenode
        self.postOrder(treenode.left)
        self.postOrder(treenode.right)
        print treenode.feature, treenode.value, treenode.maxClass

    def getNumLeafs(self, treenode):
        numLeafs = 0
        if treenode.left == 0 and treenode.right == 0:
            numLeafs += 1
        else:
            numLeafs += self.getNumLeafs(treenode.left)
            numLeafs += self.getNumLeafs(treenode.right)
        return numLeafs

    def getClass(self, x, treenode):
        if treenode.left == 0 and treenode.right == 0:
            return treenode.maxClass
        else:
            if x[treenode.feature] > treenode.value:
                return self.getClass(x, treenode.right)
            else:
                return self.getClass(x, treenode.left)

        
class cart(object):

    def __init__(self, tol = 0.0001, leastSample = 1,
                 maxDepth = inf, merge = None):
        self.tol = tol
        self.leastSample = leastSample
        self.tree = BTree()
        self.maxDepth = maxDepth
        self.merge = merge
    
    def loadData(self, data):
        dataMat = []
        for t, row in enumerate(data):
            fltRow = map(float, row)
            dataMat.append(fltRow)
        dataMat = array(dataMat)
        return dataMat

    def gini(self, y):
        ret = 0
        y_dict = {}
        for i in y:
            y_dict.setdefault(i, 0)
            y_dict[i] += 1
        for key in y_dict:
            ret += (y_dict[key]/float(len(y))) ** 2
        ret = 1 - ret
        return ret

    def classify(self, y):
        y_dict = {}
        for i in y:
            y_dict.setdefault(i, 0)
            y_dict[i] += 1
        Max = 0
        for key in y_dict:
            if y_dict[key] > Max:
                Max = y_dict[key]
                MaxValue = key
        return MaxValue

    def chooseBestSplit(self, data, y, tol, leastSample):
        tol = self.tol
        leastSample = self.leastSample
        if len(set(y)) == 1:
            return None, None, y[0]
        m, n = shape(data)
        S = self.gini(y)
        bestS = inf; bestIndex = 0; bestValue = 0
        for featIndex in range(n):
            featSet = set(data[:,featIndex])
            lastnewS = inf
            if len(featSet) == 1:
                continue
            for splitVal in featSet:
                mat0, y0, mat1, y1 = self.binSplitDataSet(data, y, featIndex, splitVal)
                if (shape(mat0)[0] < leastSample) or (shape(mat1)[0] < leastSample):
                    continue
                r0 = float(len(y0)) / len(y)
                r1 = float(len(y1)) / len(y)
                newS = r0 * self.gini(y0) + r1 * self.gini(y1)
                if newS > lastnewS:
                    break
                if newS < bestS:
                    bestIndex = featIndex
                    bestValue = splitVal
                    bestS = newS
                lastnewS = newS
        if (S - bestS) < tol:
            return None, None, self.classify(y)
        mat0, y0, mat1, y1 = self.binSplitDataSet(data, y, bestIndex, bestValue)
        if (shape(mat0)[0] < leastSample) or (shape(mat1)[0] < leastSample):
            return None, None, self.classify(y)
        return bestIndex, bestValue, self.classify(y)
        
    def createTree(self, data, y, tol, leastSample, depth = 0):
        tol = self.tol
        leastSample = self.leastSample
        retTree = TreeNode(-1, -1, -1)
        if depth >= self.maxDepth:
            retTree.feature = None
            retTree.value = None
            retTree.maxClass = self.classify(y)
            return retTree
        feat, val, mc = self.chooseBestSplit(data, y, tol, leastSample)
        if feat == None:
            retTree.feature = None
            retTree.value = None
            retTree.maxClass = mc
            return retTree
        retTree.feature = feat
        retTree.value = val
        retTree.maxClass = mc
        left, yleft, right, yright = self.binSplitDataSet(data, y, feat, val)
        retTree.left = self.createTree(left, yleft, tol, leastSample, depth + 1)
        retTree.right = self.createTree(right, yright, tol, leastSample, depth + 1)
        return retTree

    def binSplitDataSet(self, data, y, feature, value):
        determine = data[:,feature] <= value
        mat0 = data[determine]  
        mat1 = data[determine == False]
        y0 = y[determine]
        y1 = y[determine == False]
        return mat0, y0, mat1, y1

    def train(self, data, y):
        tol = self.tol
        leastSample = self.leastSample
        length = len(y)
        data = self.loadData(data)
        train = data[:(length/10)*9]
        test = data[(length/10)*9:]
        ytrain = y[:(length/10)*9]
        ytest = y[(length/10)*9:]
        if self.merge == 'pep':
            root = self.createTree(data, y, tol, leastSample)
            self.tree = BTree(root)
            root = self.prune_pep(root, data, y)
        if self.merge == 'rep':
            root = self.createTree(train, ytrain, tol, leastSample)
            root = self.prune(root, test, ytest)
        if self.merge == None:
            root = self.createTree(data, y, tol, leastSample)
        self.tree = BTree(root)

    def correctNum(self, y, y_pred):
        ret = 0
        for t, i in enumerate(y):
            if i == y_pred[t]:
                ret += 1
        return ret

    def isLeaf(self, treenode):
        return (treenode.feature == None)

    def prune(self, treenode, test, ytest):
        if not self.isLeaf(treenode.left) or not self.isLeaf(treenode.right):
            left, yleft, right, yright = self.binSplitDataSet(test, ytest,
                                            treenode.feature, treenode.value)
        if not self.isLeaf(treenode.left):
            treenode.left = self.prune(treenode.left, left, yleft)
        if not self.isLeaf(treenode.right):
            treenode.right = self.prune(treenode.right, right, yright)
        if self.isLeaf(treenode.left) and self.isLeaf(treenode.right):
            left, yleft, right, yright = self.binSplitDataSet(test, ytest,
                                            treenode.feature, treenode.value)
            yleft_pred = [treenode.left.maxClass] * len(yleft)
            yright_pred = [treenode.right.maxClass] * len(yright)
            correctNoMerge = self.correctNum(yleft_pred, yleft) + \
                           self.correctNum(yright_pred, yright)
            y_pred = [treenode.maxClass] * len(ytest)
            correctMerge = self.correctNum(y_pred, ytest)
            if correctMerge > correctNoMerge:
                print "Merging"
                treenode.feature = None
                treenode.value = None
                treenode.left = 0
                treenode.right = 0
                return treenode
            else:
                return treenode
        return treenode

    def prune_pep(self, treenode, data, y):
        if self.isLeaf(treenode):
            return treenode
        numLeafs = self.tree.getNumLeafs(treenode)
        errorNoMerge = (1 - self.score(data, y)) * len(y) + 0.5 * numLeafs
        varNoMerge = sqrt(errorNoMerge * (1 - errorNoMerge / float(len(y))))
        y_pred = [treenode.maxClass] * len(y)
        errorMerge = len(y) - self.correctNum(y_pred, y) + 0.5
        if errorMerge < errorNoMerge + sqrt(varNoMerge):
            print "Merging"
            treenode.feature = None
            treenode.value = None
            treenode.left = 0
            treenode.right = 0
            return treenode
        else:
            left, yleft, right, yright = self.binSplitDataSet(data, y,
                                            treenode.feature, treenode.value)
            if not self.isLeaf(treenode.left):
                treenode.left = self.prune_pep(treenode.left, left, yleft)
            if not self.isLeaf(treenode.right):
                treenode.right = self.prune_pep(treenode.right, right, yright)
            return treenode

    def predict(self, data):
        bt = self.tree
        ret = []
        for x in data:
            y_pred = bt.getClass(x, bt.root)
            ret.append(y_pred)
        return ret

    def score(self, data, y):
        y_pred = self.predict(data)
        return float(self.correctNum(y, y_pred)) / len(y)
        
        
if __name__ == '__main__':
    filename = "/home/yejiming/桌面/Kaggle/OttoGroup/train.csv"
    df = pd.read_csv(filename, header = 0)
    df = df.drop(['id'], axis = 1)
    data = df.values
    random.shuffle(data)
    y = data[:, -1]
    data = data[:, 0:-1]
    train = data[0:50000]
    ytrain = y[0:50000]
    test = data[50000:60000]
    ytest = y[50000:60000]
    learner = cart()
    learner.train(train, ytrain)
    

File: java/DecisionTree.java
package RandomForest;

import java.io.*;
import java.util.*;


class TreeNode {
	int feature;
	float value;
	String maxClass;
/*	Map<String, Float> probas = new HashMap<String, Float>();*/
	TreeNode left = null;
	TreeNode right = null;
	
	public TreeNode(int feature, float value, String maxClass) {
		this.feature = feature;
		this.value = value;
		this.maxClass = maxClass;
	}
	
	public TreeNode(int feature, float value, String maxClass, TreeNode left, TreeNode right) {
		this.feature = feature;
		this.value = value;
		this.maxClass = maxClass;
/*		this.probas = probas;*/
		this.left = left;
		this.right = right;
	}
}


class BTree {
	
	TreeNode root;
	
	public BTree(TreeNode root) {
		this.root = root;
	}
	
	public BTree() {}

	public void preOrder(TreeNode node) {
		
		if (node.left == null) {
			System.out.print(node.feature);
			System.out.print("   ");
			System.out.print(node.value);
			System.out.print("   ");
			System.out.println(node.maxClass);
		}
		
		else {
			System.out.print(node.feature);
			System.out.print("   ");
			System.out.print(node.value);
			System.out.print("   ");
			System.out.println(node.maxClass);
			preOrder(node.left);
			preOrder(node.right);
		}
		
	}
	
	public int getNumLeafs(TreeNode node) {
		int numLeafs = 0;
		
		if (node.left == null) {
			numLeafs += 1;
		}
		
		else {
			numLeafs += getNumLeafs(node.left);
			numLeafs += getNumLeafs(node.right);
		}
		
		return numLeafs;
	}
	
	public String getClass(float[] x, TreeNode node) {
		
		if (node.left == null) {
			return node.maxClass;
		}
		
		else {
			if (x[node.feature] > node.value)
				return getClass(x, node.right);
			else
				return getClass(x, node.left);
		}
		
	}
}


public class DecisionTree {
	
	static double tol = 0.0001;
	static int leastSample = 1;
	static int maxDepth = 999;
	static String merge = "pep";
	static int maxFeatures = 10;
	static BTree tree = new BTree();

	public static List loadData(String filename) throws IOException {
		BufferedReader in = new BufferedReader(new FileReader(filename));
		in.readLine();
		String line;
		StringBuilder sb = new StringBuilder();
		List<String> target = new ArrayList<String>();
		List<float[]> features = new ArrayList<float[]>();
		
		while ((line = in.readLine()) != null) {
			String item[] = line.split(",");
			int length = item.length - 1;
			String last = item[item.length-1];
			target.add(last);
			float feat[] = new float[length];
			
			for (int i = 0; i < length; i++) {
				feat[i] = Float.parseFloat(item[i]);
			}
			
			features.add(feat);
		}
		
		List ret = new ArrayList();
		ret.add(features);
		ret.add(target);
		return ret;
	}
	
	public DecisionTree(float tol, int leastSample, int maxDepth, String merge, int maxFeatures) {
		this.tol = tol;
		this.leastSample = leastSample;
		this.maxDepth = maxDepth;
		this.merge = merge;
		this.maxFeatures = maxFeatures;
	}
	
	public DecisionTree() {}
	
	public static float gini(List<String> y) {
		float ret = 0;
		Map<String, Float> yDict = new HashMap<String, Float>();
		
		for (String i : y) {
			if (!yDict.containsKey(i))
				yDict.put(i,  (float) 0);
			float curr = yDict.get(i);
			yDict.put(i, curr + 1);
		}
		
		for (String i : yDict.keySet()) {
			float curr = yDict.get(i);
			ret += (curr / y.size()) * (curr / y.size());
		}
		
		return 1 - ret;
	}
	
	public static String classify(List<String> y) {
		Map<String, Float> yDict = new HashMap<String, Float>();
		
		for (String i : y) {
			if (!yDict.containsKey(i))
				yDict.put(i,  (float) 0);
			float curr = yDict.get(i);
			yDict.put(i, curr + 1);
		}
		
		float Max = (float) 0;
		String maxId = null;
		
		for (String i : yDict.keySet()) {
			float curr = yDict.get(i);
			if (curr > Max) {
				Max = curr;
				maxId = i;
			}
		}
		
		return maxId;
	}
	
	public static List binSplitDataSet(List<float[]> data, List<String> y, int feature, float value) {
		List ret = new ArrayList();
		List<float[]> mat0 = new ArrayList<float[]>();
		List<float[]> mat1 = new ArrayList<float[]>();
		List<String> y0 = new ArrayList<String>();
		List<String> y1 = new ArrayList<String>();
		
		for (int i = 0; i < data.size(); i++) {
			float[] row = data.get(i);
			if (row[feature] <= value) {
				mat0.add(row);
				y0.add(y.get(i));
			}
			else {
				mat1.add(row);
				y1.add(y.get(i));
			}
		}
		
		ret.add(mat0); ret.add(y0); ret.add(mat1); ret.add(y1);
		return ret;
	}
	
	public static List chooseBestSplit(List<float[]> data, List<String> y) {
		List ret = new ArrayList();
		Set<String> ySet = new HashSet<String>(y);
		
		if (ySet.size() == 1) {
			ret.add(null); ret.add(null); ret.add(y.get(0));
			return ret;
		}
		
		int n = data.get(0).length;
		float S = gini(y);
		float bestS = (float) 1.0;
		int bestIndex = 0;
		float bestValue = (float) 0.0;
		List<Integer> featureList = new ArrayList<Integer>();
		int visitedFeatures = 0;
		
		for (int i = 0; i < n; i++)
			featureList.add(i);
		
		while (visitedFeatures < maxFeatures && featureList.size() > 0) {
			Random r = new Random();
			int randNum = r.nextInt(featureList.size());
			int featIndex = featureList.get(randNum);
			featureList.remove(randNum);
			List<Float> column = new ArrayList<Float>();
			
			for (float[] i : data) 
				column.add(i[featIndex]);
			
			Set<Float> featSet = new HashSet<Float>(column);
			if (featSet.size() == 1)
				continue;
			
			visitedFeatures += 1;
			float lastS = (float) 1.0;
			
			for (float splitVal : featSet) {
				List splitted = binSplitDataSet(data, y, featIndex, splitVal);
				List<float[]> mat0 = (ArrayList<float[]>) splitted.get(0);
				List<String> y0 = (ArrayList<String>) splitted.get(1);
				List<float[]> mat1 = (ArrayList<float[]>) splitted.get(2);
				List<String> y1 = (ArrayList<String>) splitted.get(3);
				
				if (mat0.size() < leastSample || mat1.size() < leastSample)
					continue;
				
				float r0 = ((float) y0.size()) / y.size();
				float r1 = ((float) y1.size()) / y.size();
				float newS = r0 * gini(y0) + r1 * gini(y1);
				
				if (newS > lastS)
					break;
				
				if (newS < bestS) {
					bestIndex = featIndex;
					bestValue = splitVal;
					bestS = newS;
				}
				
				lastS = newS;
			}
		}
		
		if (S - bestS < tol) {
			ret.add(null); ret.add(null); ret.add(classify(y));
			return ret;
		}
		
		List splitted = binSplitDataSet(data, y, bestIndex, bestValue);
		List<float[]> mat0 = (ArrayList<float[]>) splitted.get(0);
		List<String> y0 = (ArrayList<String>) splitted.get(1);
		List<float[]> mat1 = (ArrayList<float[]>) splitted.get(2);
		List<String> y1 = (ArrayList<String>) splitted.get(3);
		
		if (mat0.size() < leastSample || mat1.size() < leastSample) {
			ret.add(null); ret.add(null); ret.add(classify(y));
			return ret;
		}
		
		ret.add(bestIndex); ret.add(bestValue); ret.add(classify(y));
		return ret;
	}
	
	public static TreeNode createTree(List<float[]> data, List<String> y, int depth) {
		TreeNode retTree = new TreeNode(-1, -1, "Nan");
		
		if (depth > maxDepth) {
			retTree.maxClass = classify(y);
			return retTree;
		}
		
		List bestSplit = chooseBestSplit(data, y);
		Object feat = bestSplit.get(0); Object val = bestSplit.get(1); String mc = (String) bestSplit.get(2);
		if (feat == null) {
			retTree.maxClass = mc;
			return retTree;
		}
		
		retTree.feature = (int) feat;
		retTree.value = (float) val;
		retTree.maxClass = mc;
		List splitted = binSplitDataSet(data, y,  (int) feat, (float) val);
		List<float[]> left = (ArrayList<float[]>) splitted.get(0);
		List<String> yleft = (ArrayList<String>) splitted.get(1);
		List<float[]> right = (ArrayList<float[]>) splitted.get(2);
		List<String> yright = (ArrayList<String>) splitted.get(3);
		retTree.left = createTree(left, yleft, depth + 1);
		retTree.right = createTree(right, yright, depth + 1);
		
		return retTree;
	}
	
	public static boolean isLeaf(TreeNode node) {
		return (node.left == null);
	}
	
	public static TreeNode prune_pep(TreeNode node, List<float[]> data, List<String> y) {
		if (isLeaf(node))
			return node;
		
		int numLeafs = tree.getNumLeafs(node);
		long length = y.size();
		double errorNoMerge = (1 - score(data, y)) * length + 0.5 * numLeafs;
		double varNoMerge = Math.sqrt(errorNoMerge * (1 - errorNoMerge / length));
		List<String> yPred = new ArrayList<String>();
		
		for (int i = 0; i < length; i++) {
			yPred.add(node.maxClass);
		}
		
		double errorMerge = length - correctNum(y, yPred) + 0.5;
		
		if (errorMerge < errorNoMerge + varNoMerge) {
			node.feature = -1; node.value = -1; node.left = null; node.right = null;
			return node;
		}
		
		else {
			List splitted = binSplitDataSet(data, y, node.feature, node.value);
			List<float[]> left = (ArrayList<float[]>) splitted.get(0);
			List<String> yleft = (ArrayList<String>) splitted.get(1);
			List<float[]> right = (ArrayList<float[]>) splitted.get(2);
			List<String> yright = (ArrayList<String>) splitted.get(3);
			
			if (!isLeaf(node.left))
				node.left = prune_pep(node.left, left, yleft);
			
			if (!isLeaf(node.right))
				node.right = prune_pep(node.right, right, yright);
			
			return node;
		}
	}
	
	public static void train(List<float[]> data, List<String> y) {
		long length = y.size();
		long cutPoint = length * 9 / 10;
		List<String> ytrain = new ArrayList<String>();
		List<float[]> train = new ArrayList<float[]>();
		List<String> ytest = new ArrayList<String>();
		List<float[]> test = new ArrayList<float[]>();
		
		for (int i = 0; i < data.size(); i ++) {
			if (i < cutPoint) {
				train.add(data.get(i));
				ytrain.add(y.get(i));
			}
			
			else {
				test.add(data.get(i));
				ytest.add(y.get(i));
			}
		}
		
		if (merge == "None") {
			TreeNode root = createTree(data, y, 0);
			tree = new BTree(root);
		}
		
		else if (merge == "pep" ) {
			TreeNode root = createTree(data, y, 0);
			tree = new BTree(root);
			root = prune_pep(root, data, y);
			tree = new BTree(root);
		}
		
		else
			System.out.println("Prune type does not exist!");
	}
	
	public static List predict(List<float[]> data) {
		List<String> ret = new ArrayList<String>();
		
		for (float[] x : data) {
			String pred = tree.getClass(x, tree.root);
			ret.add(pred);
		}
		
		return ret;
	}
	
	public static long correctNum(List<String> y, List<String> yPred) {
		long ret = 0;
		
		for (int i = 0; i < y.size(); i++) {
			if (y.get(i).equals(yPred.get(i)))
				ret += 1;
		}
		
		return ret;
	}
	
	public static double score(List<float[]> data, List<String> y) {
		List<String> yPred = new ArrayList<String>();
		yPred = predict(data);
		return correctNum(y, yPred) / ( (double) y.size() );
	}
	
	public static void main(String[] args) throws IOException  {
		String filename = "/home/yejiming/desktop/python/ML_Algorithms/dataset.csv";
		List ret = loadData(filename);
		List<float[]> data = (List<float[]>) ret.get(0);
		List<String> y = (List<String>) ret.get(1);
		List<String> ytrain = new ArrayList<String>();
		List<float[]> train = new ArrayList<float[]>();
		List<String> ytest = new ArrayList<String>();
		List<float[]> test = new ArrayList<float[]>();
		
		for (int i = 0; i < data.size(); i ++) {
			if (i < 50000) {
				train.add(data.get(i));
				ytrain.add(y.get(i));
			}
			
			else {
				test.add(data.get(i));
				ytest.add(y.get(i));
			}
		}
		
		cart learner = new DecisionTree((float) (tol=0.0001), leastSample=1, maxDepth=999, merge="pep", maxFeatures=93);
		learner.train(train, ytrain);
		double s = learner.score(test, ytest);
/*		List<String> yPred = new ArrayList<String>();
		yPred = learner.predict(test);
		System.out.println(yPred.get(0).equals(ytest.get(0)));*/
		System.out.println(s);
	}
}
